tests/EdgeToEdge/Routes/AddDonationRouteTest.php

```diff
@@ -657,4 +658,72 @@ public function testWhenInitiallyIntendedPaymentOptionsDifferFromActual_itIsRefl
 		$this->assertContains( '12/0', $responseContent );
 	}
 
+	public function testWhenMobileTrackingIsRequested_piwikTrackerIsCalledForPaypalPayment() {
+		$this->createEnvironment( [], function ( Client $client, FunFunFactory $factory ) {
+			$factory->setNullMessenger();
+			$client->followRedirects( false );
+
+			$trackerSpy = new ServerSideTrackerSpy();
+			$factory->setServerSideTracker( $trackerSpy );
+
+			$client->request(
+				'POST',
+				'/donation/add',
+				$this->newValidMobilePayPalInput(),
+				[],
+				[ 'REMOTE_ADDR' => '10.1.2.3' ]
+			);
+
+			$client->getResponse();
+
+			$this->assertSame( '10.1.2.3', $trackerSpy->getIps()[0] );
+			$pageViews = $trackerSpy->getPageViews();
+			$this->assertCount( 1, $pageViews, 'The page view should have been tracked' );
+			// Assembled from config.test.json and request values
+			$excpectedSourceUrl = 'http://test-spenden.wikimedia.local/paypal-redir/?piwik_campaign=test&piwik_kwd=gelb';
+			$this->assertEquals( $excpectedSourceUrl, $pageViews[0]['url'] );
+		} );
+	}
+
+	private function newValidMobilePayPalInput() {
+		return [
+			'betrag' => '12,34',
+			'zahlweise' => 'PPL',
+			'periode' => 3,
+			'addressType' => 'anonym',
+			'piwik_campaign' => 'test',
+			'piwik_kwd' => 'gelb',
+			'mbt' => '1' // mobile tracking param
+		];
+	}
+
+	public function testWhenMobileTrackingIsRequested_piwikTrackerIsNotCalledForCreditCardPayment() {
+		$this->createEnvironment( [], function ( Client $client, FunFunFactory $factory ) {
+			$factory->setNullMessenger();
+			$client->followRedirects( false );
+
+			$trackerSpy = new ServerSideTrackerSpy();
+			$factory->setServerSideTracker( $trackerSpy );
+
+			$client->request(
+				'POST',
+				'/donation/add',
+				array_merge(
+					$this->newValidCreditCardInput(),
+					[ 'mbt' => '1' ]
+				)
+			);
+
+			$client->getResponse();
+
+			$this->assertCount( 0, $trackerSpy->getPageViews(), 'The page view should not have been tracked' );
+		} );
+	}
+
+	// strictly speaking, we should also write the tests
+	// testWhenMobileTrackingIsRequested_piwikTrackerIsNotCalledForDirectDebitPayment
+	// testWhenMobileTrackingIsRequested_piwikTrackerIsNotCalledForBankTransferPayment
+	// but somehow that feels ugly, especially when thinking about add more payment methods
+	// Maybe there should be a payment-specific ServersideTrackingStrategy class/factory that can be tested outside of
+	// the edge-to-edge routing test? Other ideas?
```

Comment:
Looked at the code now.

There are at least two things to test here:

* The if statement in `AddDonationHandler `that bails on non-paypal
* That `AddDonationHandler` invokes things (including the `PageViewTracker`) correctly

For the later, doing a single invocation should suffice, and is this done by the first test here. For the former we just need another test that verifies the `PageViewTracker` does not get called when non-paypal. (If it got more complicated than just that if statement, it'd be unfortunate to have it in the route handler.) That's almost what you did with the second test (it's making an assertion about `ServerSideTracker`, which I think it should not know about.)

So it seems like the spy is on the wrong level, and there is no advantage to having a `PageViewTracker` containing a `ServerSideTracker`, unless I'm missing something of course. But it looks like these can be merged.